src/Middleware/WebSockets/test/UnitTests/Http2WebSocketTests.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.WebSockets.Tests;
+
+public class Http2WebSocketTests
+{
+    [Fact]
+    public async Task Http2Handshake_Success()
+    {
+        using var host = new HostBuilder()
+            .ConfigureWebHost(webHost =>
+            {
+                webHost.UseTestServer();
+                webHost.Configure(app =>
+                {
+                    app.UseWebSockets();
+                    app.Run(httpContext =>
+                    {
+                        Assert.True(httpContext.WebSockets.IsWebSocketRequest);
+                        Assert.Equal(new[] { "p1", "p2" }, httpContext.WebSockets.WebSocketRequestedProtocols);
+                        return httpContext.WebSockets.AcceptWebSocketAsync("p2");
+                    });
+                });
+            }).Start();
+
+        var testServer = host.GetTestServer();
+
+        var result = await testServer.SendAsync(httpContext =>
+        {
+            httpContext.Request.Method = HttpMethods.Connect;
+            httpContext.Features.Set<IHttpExtendedConnectFeature>(new ConnectFeature()
+            {
+                IsExtendedConnect = true,
+                Protocol = "WebSocket",
```

Comment:
Can we test sending and receiving data frames over the WebSocket and then gracefully closing it here? I understand that nothing should change relative to HTTP/1.1 WebSockets once you get the underlying Stream, but it's good to verify.

It's also tempting to open up the stream and connection windows in both directions Http2ConnectionTests and create a duplex Stream over the data frames for the request and wrap that in a client WebSocket so we can get better end-to-end testing of this scenario since we don't have any HttpClient or JS WebSocket based tests for the end-to-end. That'd be a more work, but probably not too much.

Reply:
I'm planning to add more functional tests once the HttpClient implementation is available. Without that it's pretty hard to write any realistic tests. The client team has asked us to merge this ASAP so they can use it develop the client.